modules/core/src/smithy4s/codecs/FieldRenderPredicateCompiler.scala

```diff
@@ -0,0 +1,135 @@
+
+      def compile[A](
+          field: Field[?, A]
+      ): FieldRenderPredicateCompiler.ShouldSkip[A] = {
+        val r1 = self.compile(field)
+        val r2 = other.compile(field)
+        a => r1(a) || r2(a)
+      }
+    }
+}
+
+object FieldRenderPredicateCompiler {
+
+  type ShouldSkip[A] = A => Boolean
+
+  trait OptionalPredicate extends FieldRenderPredicateCompiler {
+
+    final def compile[A](
+        field: Field[?, A]
+    ): FieldRenderPredicateCompiler.ShouldSkip[A] = {
+      if (field.isRequired) Function.const(false)
+      else compileOptional(field)
+
+    }
+
+    def compileOptional[A](
+        field: Field[?, A]
+    ): FieldRenderPredicateCompiler.ShouldSkip[A]
+
+  }
+
+  val NeverSkip = new FieldRenderPredicateCompiler {
+    def compile[A](field: Field[_, A]): ShouldSkip[A] = Function.const(false)
+  }
+
+  private def asEmptyCollectionPredicate[F[_]](
+      schema: Schema[_]
+  ): Option[Any => Boolean] = {
+    import Schema._
+    schema match {
+      case c: CollectionSchema[f, a] =>
+        Some(collection => c.tag.isEmpty(collection.asInstanceOf[f[a]]))
+      case BijectionSchema(inner, _)  => asEmptyCollectionPredicate(inner)
+      case RefinementSchema(inner, _) => asEmptyCollectionPredicate(inner)
+      case OptionSchema(inner) =>
+        asEmptyCollectionPredicate(inner)
+          .map(predicate =>
+            wrappedCollection =>
+              wrappedCollection.asInstanceOf[Option[Any]].exists(predicate)
+          )
+      case LazySchema(_)           => None // ?
+      case _: MapSchema[k, v]      => None
+      case _: EnumerationSchema[_] => None
+      case _: StructSchema[_]      => None
+      case _: UnionSchema[_]       => None
+      case _: PrimitiveSchema[_]   => None
+    }
+  }
+
+  val SkipIfEmptyOptionalCollection =
+    new FieldRenderPredicateCompiler.OptionalPredicate {
+
+      def compileOptional[A](field: Field[?, A]): ShouldSkip[A] = {
+        asEmptyCollectionPredicate(field.schema) match {
+          case None          => Function.const(false)
+          case Some(isEmpty) => isEmpty
+        }
+      }
+    }
+
+  val SkipIfEmptyCollection = new FieldRenderPredicateCompiler {
+
+    def compile[A](field: Field[_, A]): ShouldSkip[A] = {
+      asEmptyCollectionPredicate(field.schema) match {
+        case None          => Function.const(false)
+        case Some(isEmpty) => isEmpty
+      }
+    }
+  }
+
+  val SkipIfDefaultOptionals =
+    new FieldRenderPredicateCompiler.OptionalPredicate {
+      def compileOptional[A](field: Field[?, A]): ShouldSkip[A] = {
+        // Optional fields have None as their default, so we need to make sure not to skip them here
+        a => a != None && field.isDefaultValue(a)
+      }
+    }
+
+  val SkipIfEmptyOptionals =
+    new FieldRenderPredicateCompiler.OptionalPredicate {
+      def compileOptional[A](field: Field[?, A]): ShouldSkip[A] = { a =>
+        a == None
+      }
+    }
+
+  val SkipIfEmptyOrDefaultOptionals =
```

Comment:
I'd rather we removed all the ones that are mere combinations, and let the users combine them themselves

Reply:
I agree in general but I also think that users should have an easy way to use the previous behavior and `SkipIfEmptyOrDefaultOptionals` represents the old `explicitDefaults=false`.

Reply:
isn't `explicitDefaults=false` the default?

we should have something for the current `explicitDefaults=true` though

Reply:
> isn't `explicitDefaults=false` the default?

yes, it is. 

> we should have something for the current `explicitDefaults=true` though

that is just the current `NeverSkip`